### Segmenting and Clustering Neighborhoods in Toronto

In [ ]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
!pip install folium
import folium # map rendering library

print('Libraries imported.')

In [ ]:
# To build the code to scrape the following Wikipedia page
#$ pip install beautifulsoup4
#$ easy_install beautifulsoup4
from bs4 import BeautifulSoup
url=  requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
#with open (url) as html_file:
soup = BeautifulSoup(url.text,'lxml')
#print(soup.prettify())

In [ ]:
import csv
csv_file=open('Toronto_Codes.csv','w')
csv_writer=csv.writer(csv_file)
csv_writer.writerow(['Postcode', 'Borough', 'Neighbourhood'])
for tr in soup.find_all('tr')[1:]:
    tds = tr.find_all('td')
    if len(tds)==3:
        Postcode=tds[0].text
        Borough=tds[1].text
        Neighbourhood=tds[2].text
    #print(Postcode, Borough, Neighbourhood)
    csv_writer.writerow([Postcode, Borough, Neighbourhood])
csv_file.close()

### Load table from a clean Toronto_Codes csv data

In [2]:
#load canada data from CSV
import pandas as pd
df_cd=pd.read_csv('Toronto_Codes.csv')
print('Data loaded')

Data loaded


In [3]:
df_cd.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [4]:
df_cd[df_cd.columns] = df_cd.apply(lambda x: x.str.strip('\n'))
df_cd.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### Ignore cells with a borough that is Not assigned

In [5]:
# excluding borough
df_ab=df_cd[df_cd.Borough!='Not assigned']
df_ab.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

### Get more then one neighborhood exist in one postal code area are 
like, get M5A is listed twice that has two neighborhoods: Harbourfront and Regent Park. 

In [6]:
df_ab.set_index(['Postcode','Borough'],inplace=True)
res = df_ab.groupby(level=['Postcode','Borough'], sort=False).agg( ','.join)
#res.dtypes

In [7]:
res.reset_index() # to remove set index

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
..       ...               ...   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Not assigned  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7                                      Don Mills North  
8                       Woodbine Gardens,Parkview Hill  
9                              Ryerson,Garden District  
10                                           Glencairn  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  
12                Highland Creek,Rouge Hill,Port Union  
13                     Flemingdon Park,Don Mills South  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  
18                   Guildwood]],Morningside,West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                               Caledonia-Fairbanks]]  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                                    Thorncliffe Park  
..                                                 ...  
73                                  North Toronto West  
74            

### A cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 
So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

In [8]:
res.loc[res['Neighbourhood'] =='Not assigned']=res[res['Neighbourhood'] == 'Not assigned'].index.values[0][1]
res.head()

Neighbourhood
Postcode Borough                                          
M3A      North York                              Parkwoods
M4A      North York                       Victoria Village
M5A      Downtown Toronto         Harbourfront,Regent Park
M6A      North York        Lawrence Heights,Lawrence Manor
M7A      Queen's Park                         Queen's Park

In [9]:
df=res.reset_index()
df.head()

Postcode           Borough                    Neighbourhood
0      M3A        North York                        Parkwoods
1      M4A        North York                 Victoria Village
2      M5A  Downtown Toronto         Harbourfront,Regent Park
3      M6A        North York  Lawrence Heights,Lawrence Manor
4      M7A      Queen's Park                     Queen's Park

### Using the .shape method to print the number of rows of your dataframe.

In [40]:
df.shape

(103, 6)

Now, you can see that we have 103 rows and 3 columns

In [35]:
df["Address"]=df['Borough'] +", " +df['Postcode'] 
df.head()

Postcode           Borough                    Neighbourhood  \
0      M3A        North York                        Parkwoods   
1      M4A        North York                 Victoria Village   
2      M5A  Downtown Toronto         Harbourfront,Regent Park   
3      M6A        North York  Lawrence Heights,Lawrence Manor   
4      M7A      Queen's Park                     Queen's Park   

                 Address   Latitude  Longitude  
0        North York, M3A  43.753259 -79.329656  
1        North York, M4A  43.725882 -79.315572  
2  Downtown Toronto, M5A  43.654260 -79.360636  
3        North York, M6A  43.718518 -79.464763  
4      Queen's Park, M7A  43.664366 -79.392328

### Generate list of urls using my API key first

In [49]:
url2 = [] #generate list of urls
for i in range(0, len(df)):
    url = 'https://maps.googleapis.com/maps/api/geocode/json?address={},+Toronto+Ontario,+CA&key=YOUR-KEY'.format(df['Address'][i])
    url2.append(url)

#### Now loop through the url to get latitude and longitude from google API

In [ ]:
df['Latitude']=0
df['Longitude']=0
for i, url in enumerate(url2):
    r = requests.get(url2[i], params=params)
    results = r.json()['results']
    if results:
        location = results[0]['geometry']['location']
    else:
        print ("no results")
    df['Latitude'].iloc[i] = location['lat']
    df['Longitude'].iloc[i]= location['lng']

In [41]:
df.head(50)

Postcode           Borough  \
0       M3A        North York   
1       M4A        North York   
2       M5A  Downtown Toronto   
3       M6A        North York   
4       M7A      Queen's Park   
5       M9A         Etobicoke   
6       M1B       Scarborough   
7       M3B        North York   
8       M4B         East York   
9       M5B  Downtown Toronto   
10      M6B        North York   
11      M9B         Etobicoke   
12      M1C       Scarborough   
13      M3C        North York   
14      M4C         East York   
15      M5C  Downtown Toronto   
16      M6C              York   
17      M9C         Etobicoke   
18      M1E       Scarborough   
19      M4E      East Toronto   
20      M5E  Downtown Toronto   
21      M6E              York   
22      M1G       Scarborough   
23      M4G         East York   
24      M5G  Downtown Toronto   
25      M6G  Downtown Toronto   
26      M1H       Scarborough   
27      M2H        North York   
28      M3H        North York   
29      M4H         East York   
30      M5H  Downtown Toronto   
31      M6H      West Toronto   
32      M1J       Scarborough   
33      M2J        North York   
34      M3J        North York   
35      M4J         East York   
36      M5J  Downtown Toronto   
37      M6J      West Toronto   
38      M1K       Scarborough   
39      M2K        North York   
40      M3K        North York   
41      M4K      East Toronto   
42      M5K  Downtown Toronto   
43      M6K      West Toronto   
44      M1L       Scarborough   
45      M2L        North York   
46      M3L        North York   
47      M4L      East Toronto   
48      M5L  Downtown Toronto   
49      M6L        North York   

                                        Neighbourhood                Address  \
0                                           Parkwoods        North York, M3A   
1                                    Victoria Village        North York, M4A   
2                            Harbourfront,Regent Park  Downtown Toronto, M5A   
3                     Lawrence Heights,Lawrence Manor        North York, M6A   
4                                        Queen's Park      Queen's Park, M7A   
5                                    Islington Avenue         Etobicoke, M9A   
6                                       Rouge,Malvern       Scarborough, M1B   
7                                     Don Mills North        North York, M3B   
8                      Woodbine Gardens,Parkview Hill         East York, M4B   
9                             Ryerson,Garden District  Downtown Toronto, M5B   
10                                          Glencairn        North York, M6B   
11  Cloverdale,Islington,Martin Grove,Princess Gar...         Etobicoke, M9B   
12               Highland Creek,Rouge Hill,Port Union       Scarborough, M1C   
13                    Flemingdon Park,Don Mills South        North York, M3C   
14                                   Woodbine Heights         East York, M4C   
15                                     St. James Town  Downtown Toronto, M5C   
16                                 Humewood-Cedarvale              York, M6C   
17  Bloordale Gardens,Eringate,Markland Wood,Old B...         Etobicoke, M9C   
18                  Guildwood]],Morningside,West Hill       Scarborough, M1E   
19                                        The Beaches      East Toronto, M4E   
20                                        Berczy Park  Downtown Toronto, M5E   
21                              Caledonia-Fairbanks]]              York, M6E   
22                                             Woburn       Scarborough, M1G   
23                                            Leaside         East York, M4G   
24                                 Central Bay Street  Downtown Toronto, M5G   
25                                           Christie  Downtown Toronto, M6G   
26                                          Cedarbrae       Scarborough, M1H   
27                                  Hillcrest Village        North York, M2H   

 write to csv so do not have to run the for loop again

In [ ]:
df.to_csv('Toronto_codes.csv')